[torch][segment_reduce] Add support for mean reduction (cpu)

serhaty opened the pull request:
Summary:
This diff is adding support for mean reduction for CPU (fwd + bckwd).

Will add cuda implementation in subsequent PR. We are using "cub::DeviceSegmentedReduce" for other aggregation, trying to see how to support mean or will write custom kernel for it.

Next Steps:
- cuda support for mean
- 2d data input support
- more testing
- benchmarking

Test Plan: updated unit test. Still relying on manual data for ease of debugging. Will add more tests that covers edge cases once major features are complete.

Differential Revision: D28922547

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10034,7 +10034,7 @@
     CPU, CUDA: segment_reduce_kernel
 
-- func: segment_reduce_backward(Tensor grad, Tensor output, Tensor data, *, Tensor? lengths=None) -> Tensor
+- func: segment_reduce_backward(Tensor grad, Tensor output, Tensor data, str reduce, *, Tensor? lengths=None) -> Tensor
```
This is BC breaking, must provide default

serhaty replied:
thanks for taking a look @ezyang . Afaik, currently there is no user of this function (internally or externally). Because of that, wanted to keep signature like this.
Given this, would you recommend to make it default or keep with this version?

ezyang replied:
then you must shut up the BC test by whitelisting this op. This op is in prime namespace, if you expect to be continuing to change its semantics you probably should underscore it or something.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -34,15 +35,22 @@ Tensor _segment_reduce_cpu_kernel(
           output_data[i] = initial_value;
+          if (reduction == SegmentReductionType::MEAN && lengths_data[i] > 0 &&
+              !at::_isnan(output_data[i])) {
+            output_data[i] = output_data[i] / lengths_data[i];
```
This is a different behavior from regular `torch.mean`. `torch.mean` produces `nan` when averaging over 0 elements (whereas you would produce 0)

ngimel replied:
Yep, that sounds good! (That's the same we do for max, right?)

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -116,9 +136,11 @@ Tensor segment_reduce_kernel(
   TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
   TORCH_CHECK(data.dim() == 1);
   TORCH_CHECK(data.numel() > 0);
+  auto reduction_it = at::native::segmentReduce2REDUCE.find(reduce);
```
I'm not sure using `std::map` is the best here, it's a heavyweight container. `scatter_reduce` uses just conditional statement (see `get_operator_enum`).

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -73,11 +83,16 @@ def _test_max_simple_1d(self, device, dtype, unsafe, axis):
-        self._test_max_simple_1d(device, dtype, True, 0)
-        self._test_max_simple_1d(device, dtype, True, -1)
+    def test_simple_1d(self, device, dtype):
+        reductions = ["max", "mean"]
+        for reduction in reductions:
```
```suggestion
        for reduction in ("max", "mean"):
```

ngimel reviewed (commented):
This looks good. I would prefer if we had the same `mean` behavior for segment mean as for regular tensors, but I don't know what your requirements are.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -29,20 +40,42 @@ Tensor _segment_reduce_cpu_kernel(
-              : std::numeric_limits<scalar_t>::lowest();
+          // ===== step1: initialize starting value
+          scalar_t initial_value;
+          if (reduction == SegmentReductionType::MAX) {
+            initial_value = std::numeric_limits<scalar_t>::lowest();
+          } else if (reduction == SegmentReductionType::MEAN) {
+            initial_value = 0;
+          }
+          if (initial.has_value()) {
+            initial_value = initial.value().to<scalar_t>();
+          }
```
```suggestion
          if (initial.has_value()) {
            initial_value = initial.value().to<scalar_t>();
          } else if (reduction == SegmentReductionType::MAX) {
            initial_value = std::numeric_limits<scalar_t>::lowest();
          } else if (reduction == SegmentReductionType::MEAN) {
            initial_value = 0;
          }
```

ngimel reviewed (approved):
I have a small code suggestion, otherwise looks good.

Commit:
[torch][segment_reduce] Add support for mean reduction (cpu) (#59521)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/59521

This diff is adding support for mean reduction for CPU (fwd + bckwd).

Will add cuda implementation in subsequent PR. We are using "cub::DeviceSegmentedReduce" for other aggregation, trying to see how to support mean or will write custom kernel for it.

Next Steps:
- cuda support for mean
- 2d data input support
- more testing
- benchmarking

Test Plan: updated unit test. Still relying on manual data for ease of debugging. Will add more tests that covers edge cases once major features are complete.

Reviewed By: ngimel

Differential Revision: D28922547

fbshipit-source-id: 74030e20d5467acb184ae671b7493113afb8d6a1
